### ISSUE TITLE
SPIR-V: Invalid spirv generated for liter int expression used in array lookup

### ISSUE THREAD

**Issue by tafuri** [Microsoft/DirectXShaderCompiler#957]
Compiling the following minimal shader:
```hlsl
groupshared int values[16];
void main()
{
    values[(16 - 1)] = 0;
}
```
Generates the invalid spirv expression ` OpISub %long %int_16 %int_1`. The problem is that the literal are ints, while the result is of type long.

Explicitliy casting the expression to an int makes the compiler generate valid spirv:
```hlsl
groupshared int values[16];
void main()
{
    values[int(16 - 1)] = 0;
}
```

**Comment by antiagainst** [comment]
Yep. This is a known issue. @ehsannas is fixing this with #954 and some following up pull requests.

**Comment by ehsannas** [comment]
Fixed via #954
